### PR TITLE
SNOW-749408  Add function example - batch 3

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-749408

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Add doctest examples.

### CODE REVIEW

File: src/snowflake/snowpark/functions.py [hunk 1]
```diff
@@ -486,15 +486,44 @@ def avg(e: ColumnOrName) -> Column:
...
+            |3         |
+            ------------
+            <BLANKLINE>
+            >>> df.select(count("*").alias("result")).show()
+            ------------
+            |"RESULT"  |
+            ------------
+            |5         |
+            ------------
+            <BLANKLINE>
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/functions.py:525]: the indentation does not match with other examples

File: src/snowflake/snowpark/functions.py [hunk 2]
```diff
@@ -506,6 +535,18 @@ def count(e: ColumnOrName) -> Column:
...
+
+            >>> df = session.create_dataframe([[1, 2], [1, 2], [3, None], [2, 3], [3, None], [4, None]], schema=["a", "b"])
+            >>> df.select(count_distinct(col("a"), col("b")).alias("result")).show()
+            ------------
+            |"RESULT"  |
+            ------------
+            |2         |
+            ------------
+            <BLANKLINE>
+            >>> #  The result should be 2 for {[1,2],[2,3]} since the rest are either duplicate or NULL records
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/functions.py:549]: same indentation comment

File: src/snowflake/snowpark/functions.py [hunk 3]
```diff
@@ -1382,17 +1482,36 @@ def to_date(e: ColumnOrName, fmt: Optional["Column"] = None) -> Column:
 
 
 def current_timestamp() -> Column:
-    """Returns the current timestamp for the system."""
+    """Returns the current timestamp for the system.
+
+    Example:
+    >>> import datetime
+    >>> result = session.create_dataframe([1]).select(current_timestamp()).collect()
+    >>> assert isinstance(result[0]["CURRENT_TIMESTAMP()"], datetime.datetime)
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/functions.py:1490]: indent to right

File: src/snowflake/snowpark/functions.py [hunk 4]
```diff
@@ -1382,17 +1482,36 @@ def to_date(e: ColumnOrName, fmt: Optional["Column"] = None) -> Column:
...
 
 
 def current_date() -> Column:
-    """Returns the current date for the system."""
+    """Returns the current date for the system.
+
+    Example:
+    >>> import datetime
+    >>> result = session.create_dataframe([1]).select(current_date()).collect()
+    >>> assert isinstance(result[0]["CURRENT_DATE()"], datetime.date)
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/functions.py:1502]: indent right

File: src/snowflake/snowpark/functions.py [hunk 5]
```diff
@@ -1382,17 +1482,36 @@ def to_date(e: ColumnOrName, fmt: Optional["Column"] = None) -> Column:
...
 
 
 def current_time() -> Column:
-    """Returns the current time for the system."""
+    """Returns the current time for the system.
+
+    Example:
+    >>> import datetime
+    >>> result = session.create_dataframe([1]).select(current_time()).collect()
+    >>> assert isinstance(result[0]["CURRENT_TIME()"], datetime.time)
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/functions.py:1513]: indent right

**Review by sfc-gh-aalam — APPROVED** [review]
LGTM. Only a few indentation related comments.

File: src/snowflake/snowpark/functions.py [hunk 6]
```diff
@@ -514,14 +555,36 @@ def count_distinct(*cols: ColumnOrName) -> Column:
...
-    """Returns the population covariance for non-null pairs in a group."""
+    """Returns the population covariance for non-null pairs in a group.
+
+    Example:
+        >>> df = session.create_dataframe([[1, 2], [1, 2], [4, 5], [2, 3], [3, None], [4, None], [6,4]], schema=["a", "b"])
+        >>> df.select(covar_pop(col("a"), col("b")).alias("result")).show()
+        ----------------------
+        |"RESULT"            |
+        ----------------------
+        |1.8400000000000005  |
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/functions.py:566]: Can we do a cast to only return `1.84`? I'm not sure how consistent the result would be for the 15th digit after decimal point.

File: src/snowflake/snowpark/functions.py [hunk 7]
```diff
@@ -514,14 +555,36 @@ def count_distinct(*cols: ColumnOrName) -> Column:
...
-    """Returns the sample covariance for non-null pairs in a group."""
+    """Returns the sample covariance for non-null pairs in a group.
+
+    Example:
+        >>> df = session.create_dataframe([[1, 2], [1, 2], [4, 5], [2, 3], [3, None], [4, None], [6,4]], schema=["a", "b"])
+        >>> df.select(covar_samp(col("a"), col("b")).alias("result")).show()
+        ----------------------
+        |"RESULT"            |
+        ----------------------
+        |2.3000000000000007  |
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/functions.py:584]: ditto

File: src/snowflake/snowpark/functions.py [hunk 8]
```diff
@@ -896,13 +959,37 @@ def ceil(e: ColumnOrName) -> Column:
 
 
 def cos(e: ColumnOrName) -> Column:
-    """Computes the cosine of its argument; the argument should be expressed in radians."""
+    """Computes the cosine of its argument; the argument should be expressed in radians.
+
+    Example:
+
+        >>> df = session.create_dataframe([[30]], schema=["deg"])
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/functions.py:966]: Let's try a value that gives an easier result so we don't assert at the 15th digit, say `0`, `math.pi` etc.

File: src/snowflake/snowpark/functions.py [hunk 9]
```diff
@@ -896,13 +959,37 @@ def ceil(e: ColumnOrName) -> Column:
...
     return builtin("cos")(c)
 
 
 def cosh(e: ColumnOrName) -> Column:
-    """Computes the hyperbolic cosine of its argument."""
+    """Computes the hyperbolic cosine of its argument.
+
+    Example:
+
+        >>> df = session.create_dataframe([[30]], schema=["deg"])
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/functions.py:984]: ditto

File: src/snowflake/snowpark/functions.py [hunk 10]
```diff
@@ -2886,6 +3019,26 @@ def cume_dist() -> Column:
...
+        >>> df = session.create_dataframe([[1, 2], [1, 2], [1,3], [4, 5], [2, 3], [3, 4], [4, 7], [3,7], [4,5]], schema=["a", "b"])
+        >>> df.select(cume_dist().over(Window.order_by("a")).alias("result")).show()
+        ----------------------
+        |"RESULT"            |
+        ----------------------
+        |0.3333333333333333  |
+        |0.3333333333333333  |
+        |0.3333333333333333  |
+        |0.4444444444444444  |
+        |0.6666666666666666  |
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/functions.py:3035]: I'm surprised that this does not end with `7` lol. I don't have a straightforward way to make this result not use long float number :/

**Review by sfc-gh-sfan — APPROVED** [review]

**Pull request merged**